File: src/access_token/access_token.service.ts
import { Injectable } from '@nestjs/common';
import { AccessToken, Prisma, PrismaClient, User } from '@prisma/client';
import { nanoid } from 'nanoid';

@Injectable()
export class AccessTokenService {
  constructor(private readonly prisma: PrismaClient) {}

  async authRegister(phone: string): Promise<AccessToken> {
    const user = await this.autoCreateUser(phone);

    return this.create(user);
  }

  async create(user: string | User): Promise<AccessToken> {
    const userId = typeof user === 'string' ? user : user.id;
    const { expiredIn, refreshIn } = await this.getOptions();
    const result = await this.prisma.accessToken.create({
      data: {
        userId,
        token: nanoid(128),
        expiredAt: new Date(Date.now() + expiredIn * 1000),
        refreshExpiredAt: new Date(Date.now() + refreshIn * 1000),
      },
    });

    this.deleteAllExpired();

    return result;
  }

  async refresh(accessToken: AccessToken): Promise<AccessToken> {
    // 当前 Token 设置为五分钟后过期
    // 五分钟之内还能使用，但是不能刷新
    await this.prisma.accessToken.update({
      where: { token: accessToken.token },
      data: {
        expiredAt: new Date(Date.now() + 60 * 5 * 1000),
        refreshExpiredAt: new Date(),
      },
    });

    return this.create(accessToken.userId);
  }

  async delete(accessToken: AccessToken): Promise<void> {
    await this.prisma.accessToken.delete({
      where: { token: accessToken.token },
    });

    await this.deleteAllExpired();
  }

  private async deleteAllExpired(): Promise<void> {
    await this.prisma.accessToken.deleteMany({
      where: {
        expiredAt: { lt: new Date() },
        refreshExpiredAt: { lt: new Date() },
      },
    });
  }

  private async autoCreateUser(phone: string): Promise<User> {
    const user = await this.prisma.user.findUnique({
      where: { phone },
      rejectOnNotFound: false,
    });

    if (user) {
      return user;
    }

    return this.prisma.user.create({
      data: {
        phone,
        id: nanoid(64),
      },
    });
  }

  private async getOptions(): Promise<{
    expiredIn: number;
    refreshIn: number;
  }> {
    const row = await this.prisma.configuration.findUnique({
      where: {
        key: 'access-token-options',
      },
      rejectOnNotFound: false,
    });
    const options: Prisma.JsonObject | null | undefined =
      row?.value as Prisma.JsonObject;
    const expiredIn = (options?.expiredIn as number) ?? 60 * 60;
    const refreshIn = (options?.refreshIn as number) ?? 60 * 60 * 24 * 7;

    return { expiredIn, refreshIn };
  }
}

File: lib/services/access_token.service.dart
// Copyright (c) 2021, Odroe Inc. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

import 'dart:io';

import 'package:grpc/grpc.dart';
import 'package:server/auth.dart';
import 'package:server/configuration.dart';
import 'package:server/database/connection_pool.dart';
import 'package:server/helpers/string.helper.dart';
import 'package:server/protos/access_token.pbgrpc.dart';
import 'package:server/protos/google/protobuf/empty.pb.dart';
import 'package:server/protos/google/protobuf/timestamp.pb.dart';
import 'package:single/single.dart';

class AccessTokenService extends AccessTokenServiceBase {
  @override
  Future<AccessTokenResponse> create(
      ServiceCall call, AccessTokenCreateRequest request) async {
    final done = await _validateOtp(request.phone, request.otp);
    final options = single<Configuration>().accessToken;
    final user = await _resolveUserFromPhone(request.phone);

    final now = DateTime.now();
    final response = AccessTokenResponse();
    response.userId = user['id'];
    response.token = StringHelper.string(128);
    response.expiredAt =
        Timestamp.fromDateTime(now.add(Duration(days: options.expiredInDays)));
    response.refreshExpiredAt = Timestamp.fromDateTime(
        now.add(Duration(days: options.refreshExpiredInDays)));
    response.createdAt = Timestamp.fromDateTime(now);

    final database = await single<DatabaseConnectionPool>().getConnection();
    database.execute(
      r'INSERT INTO access_tokens (user_id, token, expired_at, refresh_expired_at) VALUES (@userId, @token, @expiredAt, @refreshExpiredAt)',
      substitutionValues: {
        'userId': response.userId,
        'token': response.token,
        'expiredAt': response.expiredAt.toDateTime(),
        'refreshExpiredAt': response.refreshExpiredAt.toDateTime(),
      },
    );
    done.call();
    _deleteAllExpiredTokens();

    return response;
  }

  @override
  Future<AccessTokenResponse> refresh(ServiceCall call, Empty request) async {
    final accessToken =
        await single<Auth>().getAccessToken(call.clientMetadata);
    single<Auth>().validate(
      accessToken: accessToken,
      isRefresh: true,
    );

    final database = await single<DatabaseConnectionPool>().getConnection();

    // Set current access token refresh expired at to now + 5 minutes
    // and set expired at to now.
    database.execute(
      r'UPDATE access_tokens SET refresh_expired_at = NOW(), expired_at = @expiredAt, has_refreshed = true WHERE token = @token',
      substitutionValues: {
        'token': accessToken!.token,
        'expiredAt': DateTime.now().add(Duration(minutes: 5)),
      },
    );

    accessToken.token = StringHelper.string(128);
    accessToken.expiredAt = Timestamp.fromDateTime(DateTime.now().add(
        Duration(days: single<Configuration>().accessToken.expiredInDays)));
    accessToken.refreshExpiredAt = Timestamp.fromDateTime(DateTime.now().add(
        Duration(
            days: single<Configuration>().accessToken.refreshExpiredInDays)));

    /// Create new access token
    database.execute(
      r'INSERT INTO access_tokens (user_id, token, expired_at, refresh_expired_at) VALUES (@userId, @token, @expiredAt, @refreshExpiredAt)',
      substitutionValues: {
        'userId': accessToken.userId,
        'token': accessToken.token,
        'expiredAt': accessToken.expiredAt.toDateTime(),
        'refreshExpiredAt': accessToken.refreshExpiredAt.toDateTime(),
      },
    );

    _deleteAllExpiredTokens();

    return accessToken;
  }

  @override
  Future<Empty> revoke(ServiceCall call, Empty request) async {
    final accessToken =
        await single<Auth>().getAccessToken(call.clientMetadata);
    single<Auth>().validate(
      accessToken: accessToken,
      isRefresh: true,
    );

    final database = await single<DatabaseConnectionPool>().getConnection();

    // Delete access token
    database.execute(
      r'DELETE FROM access_tokens WHERE token = @token',
      substitutionValues: {
        'token': accessToken!.token,
      },
    );

    _deleteAllExpiredTokens();

    return Empty();
  }

  Future<Map<String, dynamic>> _resolveUserFromPhone(String phone) async {
    final Map<String, dynamic>? user = await _findUserFromPhone(phone);

    if (user != null) {
      return user;
    }

    final Map<String, dynamic> values = {
      'id': StringHelper.string(64),
      'phone': phone,
      'name': StringHelper.string(64),
    };
    final database = await single<DatabaseConnectionPool>().getConnection();
    await database.execute(
      r'INSERT INTO users (id, name, phone) VALUES (@id, @name, @phone)',
      substitutionValues: values,
    );

    return values;
  }

  Future<Map<String, dynamic>?> _findUserFromPhone(String phone) async {
    final database = await single<DatabaseConnectionPool>().getConnection();
    final results = await database.mappedResultsQuery(
      r'SELECT id, phone FROM users WHERE phone = @phone',
      substitutionValues: {'phone': phone},
    );

    if (results.isEmpty) {
      return null;
    }

    final row = results.first;
    if (row['users'] != null) {
      return row['users'];
    }
  }

  Future<void Function()> _validateOtp(String phone, String otp) async {
    final database = await single<DatabaseConnectionPool>().getConnection();
    final results = await database.mappedResultsQuery(
      r'SELECT expired_at FROM verification_codes WHERE phone = @phone AND code = @otp AND expired_at > NOW()',
      substitutionValues: {'phone': phone, 'otp': otp},
    );
    database.execute(
      r'DELETE FROM verification_codes WHERE expired_at < NOW()',
    );

    if (results.isEmpty) {
      throw GrpcError.invalidArgument('Invalid otp');
    }

    return () async {
      await database.execute(
        r'DELETE FROM verification_codes WHERE phone = @phone AND code = @otp',
        substitutionValues: {'phone': phone, 'otp': otp},
      );
    };
  }

  void _deleteAllExpiredTokens() async {
    final database = await single<DatabaseConnectionPool>().getConnection();
    await database.execute(
      r'DELETE FROM access_tokens WHERE expired_at < NOW() AND has_refreshed IS true',
    );
  }
}
